server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSReceiveOptionsCheckUtility.java

```diff
@@ -63,90 +64,47 @@ private void checkGroupOne(CICSParser.Cics_receive_group_oneContext ctx) {
     if (!ctx.ASIS().isEmpty() || !ctx.BUFFER().isEmpty()) {
       checkHasIllegalOptions(ctx.LEAVEKB(), "LEAVEKB");
     }
-
-    List<RuleContextData> contexts = new ArrayList<>();
-    contexts.add(new RuleContextData(ctx.cics_into_set(), "INTO or SET"));
-    contexts.add(new RuleContextData(ctx.cics_maxlength(), "MAXLENGTH or MAXFLENGTH"));
-    contexts.add(new RuleContextData(ctx.ASIS(), "ASIS", ErrorSeverity.WARNING));
-    contexts.add(new RuleContextData(ctx.BUFFER(), "BUFFER", ErrorSeverity.WARNING));
-    contexts.add(new RuleContextData(ctx.NOTRUNCATE(), "NOTRUNCATE", ErrorSeverity.WARNING));
-    contexts.add(new RuleContextData(ctx.LEAVEKB(), "LEAVEKB", ErrorSeverity.WARNING));
-    contexts.add(new RuleContextData(ctx.cics_handle_response(), "RESPONSE HANDLER"));
-    harvestResponseHandlers(ctx.cics_handle_response(), contexts);
-    checkDuplicates(contexts);
+    checkResponseHandlers(ctx.cics_handle_response());
+    checkDuplicates(ctx, "ASIS", "BUFFER", "NOTRUNCATE", "LEAVEKB");
   }
 
   private void checkGroupTwo(CICSParser.Cics_receive_group_twoContext ctx) {
     checkHasMandatoryOptions(ctx.cics_length_flength(), ctx, "LENGTH or FLENGTH");
     checkHasMandatoryOptions(ctx.cics_into_set(), ctx, "INTO or SET");
 
-    List<RuleContextData> contexts = new ArrayList<>();
-    contexts.add(new RuleContextData(ctx.CONVID(), "CONVID"));
-    contexts.add(new RuleContextData(ctx.SESSION(), "SESSION"));
-    contexts.add(new RuleContextData(ctx.cics_into_set(), "INTO or SET"));
-    contexts.add(new RuleContextData(ctx.cics_length_flength(), "LENGTH or FLENGTH"));
-    contexts.add(new RuleContextData(ctx.cics_maxlength(), "MAXLENGTH or MAXFLENGTH"));
-    contexts.add(new RuleContextData(ctx.NOTRUNCATE(), "NOTRUNCATE", ErrorSeverity.WARNING));
-    contexts.add(new RuleContextData(ctx.STATE(), "STATE"));
-    contexts.add(new RuleContextData(ctx.cics_handle_response(), "RESPONSE HANDLER"));
-    harvestResponseHandlers(ctx.cics_handle_response(), contexts);
-    checkDuplicates(contexts);
+    checkResponseHandlers(ctx.cics_handle_response());
+    Map<String, ErrorSeverity> specialSeverities = new HashMap<>();
+    specialSeverities.put("NOTRUNCATE", ErrorSeverity.WARNING);
+    checkDuplicates(ctx, "NOTRUNCATE");
```

Comment:
The biggest problem with this approach is that we lose the ability to check the logical groups e.g. "INTO or SET".

Maybe it could be addressed by extending a map I talked about in the constext of `checkDuplicateEntries`.
It could map `token_id` to `{group_id, severity}` (by default group_id is the token text, severity is error).

The `Map<Integer, TerminalNode> entries` would be replaced just by `Set<String>` String being the group id e.g. "INTO or SET" or the token text.